crates/bevy_core_pipeline/src/deferred/node.rs

```diff
@@ -0,0 +1,192 @@
+use bevy_ecs::prelude::*;
+use bevy_ecs::query::QueryItem;
+use bevy_render::render_graph::ViewNode;
+use bevy_render::view::Msaa;
+use bevy_render::{
+    camera::ExtractedCamera,
+    prelude::Color,
+    render_graph::{NodeRunError, RenderGraphContext},
+    render_phase::RenderPhase,
+    render_resource::{
+        LoadOp, Operations, RenderPassColorAttachment, RenderPassDepthStencilAttachment,
+        RenderPassDescriptor,
+    },
+    renderer::RenderContext,
+    view::ViewDepthTexture,
+};
+#[cfg(feature = "trace")]
+use bevy_utils::tracing::info_span;
+
+use crate::core_3d::{Camera3d, Camera3dDepthLoadOp};
+use crate::prepass::{DepthPrepass, MotionVectorPrepass, NormalPrepass, ViewPrepassTextures};
+
+use super::{AlphaMask3dDeferred, Opaque3dDeferred};
+
+/// Render node used by the prepass.
+///
+/// By default, inserted before the main pass in the render graph.
+#[derive(Default)]
+pub struct DeferredNode;
+
+impl ViewNode for DeferredNode {
+    type ViewQuery = (
+        &'static ExtractedCamera,
+        &'static RenderPhase<Opaque3dDeferred>,
+        &'static RenderPhase<AlphaMask3dDeferred>,
+        &'static ViewDepthTexture,
+        &'static ViewPrepassTextures,
+        &'static Camera3d,
+        Option<&'static DepthPrepass>,
+        Option<&'static NormalPrepass>,
+        Option<&'static MotionVectorPrepass>,
+    );
+
+    fn run(
+        &self,
+        graph: &mut RenderGraphContext,
+        render_context: &mut RenderContext,
+        (
+            camera,
+            opaque_deferred_phase,
+            alpha_mask_deferred_phase,
+            view_depth_texture,
+            view_prepass_textures,
+            camera_3d,
+            depth_prepass,
+            normal_prepass,
+            motion_vector_prepass,
+        ): QueryItem<Self::ViewQuery>,
+        world: &World,
+    ) -> Result<(), NodeRunError> {
+        let view_entity = graph.view_entity();
+
+        if let Some(msaa) = world.get_resource::<Msaa>() {
+            match msaa {
+                Msaa::Off => (),
+                _ => panic!("MSAA not supported when using deferred rendering."),
```

Comment:
I don't love this being a panic, but I'm not sure if disabling msaa and just logging an error would be acceptable.

Reply:
I don't love it either. Maybe it would be better to disable MSAA. I'll have to see what happens when someone adds deferred to the camera in the middle of rendering. Maybe the plugin needs to disable MSAA, *and* the panic should remain too.

Reply:
I feel like I prefer warning (ideally once) and then turning MSAA off automatically is the friendlier approach.